### PR TITLE
Added JsonNetObjectContractProvider to enable private non default constructor resolution for Newtonsoft Json.NET

### PR DESCRIPTION
**Pull request by oskardudycz**
- Added `JsonNetObjectContractProvider` to enable private non default constructor resolution for Newtonsoft Json.NET
- Added `NonPublicMembersStorage.NonPublicConstructor` option and updated `JsonNetContractResolver` accordingly
- Added tests and updated docs

### CODE REVIEW

File: src/Marten/ISerializer.cs [hunk 1]
```diff
@@ -165,6 +165,8 @@ public enum NonPublicMembersStorage
     {
         Default = 0,
         NonPublicSetters = 1,
-        NonPublicDefaultConstructor = 2
+        NonPublicDefaultConstructor = 2,
+        NonPublicConstructor = 4,
```
Comment (mysticmind) [inline src/Marten/ISerializer.cs:169]: Minor thing, shouldn't the value be `3` rather than `4`?

**Review by mysticmind — APPROVED** [review]
Looks good to me 👍

File: src/Marten/ISerializer.cs (hunk 1 above)
Comment (oskardudycz) [reply src/Marten/ISerializer.cs:169]: I'm adding that to allow bit flag usage (e.g. `NonPublicMembersStorage.NonPublicSetters | NonPublicMembersStorage.NonPublicDefaultConstructor`). It needs to be 2^n

Comment (oskardudycz) [reply src/Marten/ISerializer.cs:169]: See also: https://github.com/JasperFx/marten/pull/2004/files#diff-f899e2a4569989d2b4fbf225960c0a07bba52c31d6b434b964ffb02dfa4eb148R33

Comment (mysticmind) [reply src/Marten/ISerializer.cs:169]: Okay, makes sense 👍

**Pull request merged**